### ISSUE TITLE
Mac permissions prevent use of getSiteDataDir

### ISSUE THREAD

**Issue by mrlimbic** [harawata/appdirs#30]
Java apps don't automatically have permission to write to /Library/Application Support/

It may be better to return /Users/Shared/ or /Users/Shared/Library/Application Support/ which automatically has permissions for all users.

**Comment by harawata** [comment]
Hello @mrlimbic ,

Thank you for the suggestion!

On Linux, `/usr/local/share` is owned by `root` and its permission is `755`.
On Windows, any user can create a file in the `ProgramData` folder, but only the creator can update the file [1].

So, `/Library/Application Support` seems to be more appropriate in terms of consistency.
I am open to further discussion, though.

[1] https://stackoverflow.com/a/22107884

**Comment by mrlimbic** [comment]
It's just not very practical to use it. 

A java desktop app on Mac will not have permissions to use /Library (but it will for ~/Library). For something global that has permissions better is /Users/Shared/Library/Application Support/.

I can't distribute my app if it uses /Library. It will only work for people who are willing to run app as administrator using sudo via terminal. Nobody expects to have to do that for a Mac GUI desktop app they downloaded.

I've overriden behaviour in my code on Mac to use  /Users/Shared/Library/Application Support/ which works fine for global app settings.

**Comment by harawata** [comment]
Thank you for the follow-up, @mrlimbic ,

You want a directory to which all users have write access.
Although it is a reasonable request, it seems to differ from the concept of 'site directories' where superuser or owner has write access while other users have read-only access.

After Googling a bit, I am thinking about introducing a new method `getCommonDir()` or `getAllUsersDir()` (any better name?) which uses the following directories as the base.

- macOS : `/Users/Shared/Library/Application Support/`
- WIndows : [CSIDL_COMMON_APPDATA](https://docs.microsoft.com/en-us/windows/win32/shell/csidl#csidl_common_appdata) (=`C:\ProgramData`)
- Linux : [/srv](http://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s17.html)

Let me know what you think!

**Comment by mrlimbic** [comment]
Yes adding a new one makes sense when storing global settings for a desktop app. That doesn't fit the site dir concept either. You'll find almost no desktop apps on Mac use /Library/Application Support/ because it is so painful.

I think getAllUsersDir or getSharedDir sounds most logical for the new directory.

**Comment by harawata** [comment]
Hi @mrlimbic ,

Added `getSharedDir()` (via #31 ).
I plan to release the next version 1.1.0 within a few days, but it would be great if you could verify the new behavior using the latest [1.1.0-SNAPSHOT](https://oss.sonatype.org/content/repositories/snapshots/net/harawata/appdirs/).

If you use Maven, add the snapshot repository to the pom.xml...

```xml
<project>
...
  <repositories>
    <repository>
      <id>sonatype-snapshot</id>
      <name>Sonatype OSS Snapshot Repository</name>
      <url>http://oss.sonatype.org/content/repositories/snapshots</url>
    </repository>
  </repositories>
...
</project>
```
...and specify `1.1.0-SNAPSHOT` as appdirs version.

```xml
<dependency>
  <groupId>net.harawata</groupId>
  <artifactId>appdirs</artifactId>
  <version>1.1.0-SNAPSHOT</version>
</dependency>
```

Thanks again for taking the time to submit the issue!

**Comment by mrlimbic** [comment]
Both user directory and shared directory seem to work fine on Mac OS now.

However on Windows 10 (running in a VirtualBox VM on my Mac) I get the following error when trying to access both directories.

Is this because Windows 10 and the JDK are 64 bit?

```
Exception in thread "main" java.lang.AbstractMethodError: com.sun.jna.Structure.getFieldOrder()Ljava/util/List;
	at com.sun.jna.Structure.fieldOrder(Structure.java:952)
	at com.sun.jna.Structure.getFields(Structure.java:1006)
	at com.sun.jna.Structure.deriveLayout(Structure.java:1172)
	at com.sun.jna.Structure.calculateSize(Structure.java:1097)
	at com.sun.jna.Structure.calculateSize(Structure.java:1049)
	at com.sun.jna.Structure.allocateMemory(Structure.java:403)
	at com.sun.jna.Structure.<init>(Structure.java:194)
	at com.sun.jna.Structure.<init>(Structure.java:182)
	at com.sun.jna.Structure.<init>(Structure.java:169)
	at com.sun.jna.Structure.<init>(Structure.java:161)
	at com.sun.jna.platform.win32.Guid$GUID.<init>(Guid.java:126)
	at com.sun.jna.platform.win32.Guid$GUID.fromString(Guid.java:257)
	at com.sun.jna.platform.win32.KnownFolders.<clinit>(KnownFolders.java:41)
	at net.harawata.appdirs.impl.ShellFolderResolver.convertFolderIdToGuid(ShellFolderResolver.java:61)
	at net.harawata.appdirs.impl.ShellFolderResolver.resolveFolder(ShellFolderResolver.java:36)
	at net.harawata.appdirs.impl.WindowsAppDirs.getLocalAppData(WindowsAppDirs.java:75)
	at net.harawata.appdirs.impl.WindowsAppDirs.getUserDataDir(WindowsAppDirs.java:33)
	at net.harawata.appdirs.AppDirs.getUserDataDir(AppDirs.java:20)
	at net.vordio.util.ComputerInfo.getUserDataDir(ComputerInfo.java:44)
	at net.vordio.util.ComputerInfo.main(ComputerInfo.java:31)

```

**Comment by harawata** [comment]
That error usually occurs when there are older versions of JNA and/or JNA-Platform JAR(s) on the classpath (AppDirs 1.1.0 is built against version 5.5.0 of JNA.).

If you are using maven, it is possible that there is another dependency that pulls the older JNA versions.
If this is the case, try adding the JNA dependencies directly to your pom.xml.

```xml
<dependency>
    <groupId>net.java.dev.jna</groupId>
    <artifactId>jna</artifactId>
    <version>5.5.0</version>
</dependency>
<dependency>
    <groupId>net.java.dev.jna</groupId>
    <artifactId>jna-platform</artifactId>
    <version>5.5.0</version>
</dependency>
```

**Comment by mrlimbic** [comment]
Yes those adding maven dependencies fixed it. 

However I don't understand why did maven not do that automatically if appdirs depends on it?

**Comment by harawata** [comment]
Glad to know it's resolved!

Most likely, older versions of JNA jars remained in the 'target' directory.
Executing `mvn clean` usually resolves this issue.

**Comment by harawata** [comment]
Released [appdirs-1.1.0](https://github.com/harawata/appdirs/releases/tag/appdirs-1.1.0).
It's also available on Maven central repository.
